File: HashTableCpp/HashTableCpp/HashTable.cpp
#include <iostream>
#include <list>

#include "HashTable.hpp"

bool HashTable::__checkPrime(int value) {
	if (value <= 1) {
		return false;
	}
	for (int i = 2; i < value / 2; i++) {
		if (value % i == 0) {
			return false;
		}
	}
	return true;
}

int HashTable::__getPrime(int value) {
	while (!__checkPrime(value)) {
		value++;
	}
	return value;
}

int HashTable::__hashFunction(int key) {
	return(key % __capacity);
}

HashTable::HashTable(int value) {
	__capacity = __getPrime(value);
	__table = new std::list<std::pair<int, int>>[__capacity];
};

HashTable::~HashTable() {
	delete[] __table;
}

void HashTable::insertIntoTable(int key, int data) {
	int index = __hashFunction(key);
	__table[index].push_back(std::make_pair(key, data));
}

void HashTable::deleteFromTable(int key) {
	int index = __hashFunction(key);

	for (auto i = __table[index].begin(); i != __table[index].end();) {
		if (i->first == key) {
			__table[index].erase(i);
			break;
		}
		else {
			++i;
		}
	}
}

bool HashTable::searchKey(int key) {
	int index = __hashFunction(key);

	for (auto i = __table[index].begin(); i != __table[index].end();) {
		if (i->first == key) {
			return true;
		}
		else {
			++i;
		}
	}

	return false;

}

void HashTable::displayTable() {
	std::cout << "Your table is:";
	for (int i = 0; i < __capacity; i++) {
		std::cout << "\nTable[" << i << "]: ";
		bool skip = true;
		if (__table[i].empty()) {
			std::cout << "NULL";
		}
		else {
			std::cout << __table[i].front().second;
		}
		for (auto x : __table[i]) {
			if (skip) {
				skip = false;
				continue;
			}
			std::cout << "->" << x.second;
		}
		std::cout << std::endl;
			
	}
}

File: HashTablePython/HashTable.py
class HashTable:
    def __check_prime(self, n):
        if n <= 1: return False
        for i in range(2, int(n / 2)):
            if n % i == 0:
                return False
        return True
    
    def __get_prime(self, n):
        while self.__check_prime(n) is not True:
            n += 1
        return n

    def __hash_function(self, key):
        return (key % self.__capacity)

    def __init__(self, size):
        self.__capacity = self.__get_prime(size)
        self.__table = [[] for _ in range(self.__capacity)]

    def insert_into_table(self, key, data):
        index = self.__hash_function(key)
        if self.__table[index] == []:
            self.__table[index] = [key, data]
            return
        else:
            self.insert_into_table(key + 1, data)

    def delete_from_table(self, key_to_del):
        index = self.__hash_function(key_to_del)
        if self.__table[index][0] == key_to_del:
            self.__table[index] = []
        else:    
            print("This Key is not found!\n")


    def search_key(self, key_to_find):
        index = self.__hash_function(key_to_find)
        if self.__table[index][0] == key_to_find:
            return self.__table[index][1]
        else:
            print("This Key is not found!")

    def display_table(self):
        print("This is your Hash Table: \n", end="")
        for elem in self.__table:
            if elem == []:
                print("NIL\n", end="")
            else:
                print(f"{elem[1]}\n", end="")
        print("\n")